TESTS/block_device/qspif/main.cpp

```diff
@@ -0,0 +1,368 @@
+        basic_erase_program_read_test(blockD, block_size, write_block, read_block, error_mask, addrwidth);
+    }
+
+    err = blockD.deinit();
+    TEST_ASSERT_EQUAL(0, err);
+
+end:
+    delete[] write_block;
+    delete[] read_block;
+    delete[] error_mask;
+
+}
+
+void test_qspif_unaligned_program()
+{
+
+    printf("\nTest Unaligned Program Starts..\n");
+
+    QSPIFBlockDevice blockD(MBED_CONF_QSPIF_QSPI_IO0, MBED_CONF_QSPIF_QSPI_IO1, MBED_CONF_QSPIF_QSPI_IO2,
+                            MBED_CONF_QSPIF_QSPI_IO3, MBED_CONF_QSPIF_QSPI_CLK, MBED_CONF_QSPIF_QSPI_CS, QSPIF_POLARITY_MODE_0,
+                            MBED_CONF_QSPIF_QSPI_FREQ);
+
+    int err = blockD.init();
+    TEST_ASSERT_EQUAL(0, err);
+
+    for (unsigned atr = 0; atr < sizeof(ATTRS) / sizeof(ATTRS[0]); atr++) {
+        static const char *prefixes[] = {"", "k", "M", "G"};
+        for (int i_ind = 3; i_ind >= 0; i_ind--) {
+            bd_size_t size = (blockD.*ATTRS[atr].method)();
+            if (size >= (1ULL << 10 * i_ind)) {
+                printf("%s: %llu%sbytes (%llubytes)\n",
+                       ATTRS[atr].name, size >> 10 * i_ind, prefixes[i_ind], size);
+                break;
+            }
+        }
+    }
+
+    bd_size_t block_size = blockD.get_erase_size();
+    unsigned addrwidth = ceil(log(float(blockD.size() - 1)) / log(float(16))) + 1;
+
+    uint8_t *write_block = new (std::nothrow) uint8_t[block_size];
+    uint8_t *read_block = new (std::nothrow) uint8_t[block_size];
+    uint8_t *error_mask = new (std::nothrow) uint8_t[TEST_ERROR_MASK];
+    if (!write_block || !read_block || !error_mask) {
+        printf("\n Not enough memory for test");
+        goto end;
+    }
+
+    {
+        bd_addr_t block = (rand() * block_size) % blockD.size() + 15;
+
+        // Use next random number as temporary seed to keep
+        // the address progressing in the pseudorandom sequence
+        unsigned seed = rand();
+
+        // Fill with random sequence
+        srand(seed);
+        for (bd_size_t i_ind = 0; i_ind < block_size; i_ind++) {
+            write_block[i_ind] = 0xff & rand();
+        }
+
+        // Write, sync, and read the block
+        printf("\ntest  %0*llx:%llu...", addrwidth, block, block_size);
+
+        err = blockD.erase(block, block_size);
+        TEST_ASSERT_EQUAL(0, err);
+
+        err = blockD.program(write_block, block, block_size);
+        TEST_ASSERT_EQUAL(0, err);
+
+        printf("\nwrite %0*llx:%llu ", addrwidth, block, block_size);
+        for (int i_ind = 0; i_ind < 15; i_ind++) {
+            printf("%02x", write_block[block_size - 16 + i_ind]);
+        }
+        printf("...\n");
+
+        err = blockD.read(read_block, block, block_size);
+        TEST_ASSERT_EQUAL(0, err);
+
+        printf("read  %0*llx:%llu ", addrwidth, block, block_size);
+        for (int i_ind = 0; i_ind < 15; i_ind++) {
+            printf("%02x", read_block[block_size - 16 + i_ind]);
+        }
+        printf("...\n");
```

Comment:
Test should minimize debug output by default (these debug msg are important but only if something in the test goes wrong). Instead of printf, use `utest_printf`

Reply:
Will fix - thanks